scylla-cql/src/types/deserialize/value.rs

```diff
@@ -0,0 +1,1515 @@
+                );
+                Ok(ret)
+            }
+        }
+    }
+}
+
+macro_rules! impl_tuple_multiple {
+    (;;) => {
+        impl_tuple!(;;);
+    };
+    ($TN:ident $(,$Ti:ident)*; $idx_n:literal $(,$idx:literal)*; $idf_n:ident $(,$idf:ident)*) => {
+        impl_tuple_multiple!($($Ti),*; $($idx),*; $($idf),*);
+        impl_tuple!($TN $(,$Ti)*; $idx_n $(,$idx)*; $idf_n $(,$idf)*);
+    }
+}
+
+impl_tuple_multiple!(
+    T0, T1, T2, T3, T4, T5, T6, T7, T8, T9, T10, T11, T12, T13, T14, T15;
+    0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15;
+    t0, t1, t2, t3, t4, t5, t6, t7, t8, t9, t10, t11, t12, t13, t14, t15
+);
+
+// udts
+
+/// An iterator over fields of a User Defined Type.
+///
+/// # Note
+///
+/// A serialized UDT will generally have one value for each field, but it is
+/// allowed to have less. This iterator differentiates null values
+/// from non-existent values in the following way:
+///
+/// - `None` - missing from the serialized form
+/// - `Some(None)` - present, but null
+/// - `Some(Some(...))` - non-null, present value
+pub struct UdtIterator<'frame> {
+    fields: &'frame [(String, ColumnType)],
+    raw_iter: BytesSequenceIterator<'frame>,
+}
+
+impl<'frame> UdtIterator<'frame> {
+    #[inline]
+    pub fn new(fields: &'frame [(String, ColumnType)], slice: FrameSlice<'frame>) -> Self {
+        Self {
+            fields,
+            raw_iter: BytesSequenceIterator::new(slice),
+        }
+    }
+
+    #[inline]
+    pub fn fields(&self) -> &'frame [(String, ColumnType)] {
+        self.fields
+    }
+}
+
+impl<'frame> DeserializeCql<'frame> for UdtIterator<'frame> {
+    fn type_check(typ: &ColumnType) -> Result<(), ParseError> {
+        match typ {
+            ColumnType::UserDefinedType { .. } => Ok(()),
+            _ => Err(ParseError::BadIncomingData(format!(
+                "Expected a user defined type, got {:?}",
+                typ,
+            ))),
+        }
+    }
+
+    fn deserialize(
+        typ: &'frame ColumnType,
+        v: Option<FrameSlice<'frame>>,
+    ) -> Result<Self, ParseError> {
+        let v = ensure_not_null_slice(v)?;
+        let mem = v.as_slice();
+        let fields = match typ {
+            ColumnType::UserDefinedType { field_types, .. } => field_types.as_ref(),
+            _ => {
+                return Err(ParseError::BadIncomingData(format!(
+                    "Expected a user defined type, got {:?}",
+                    typ,
+                )))
+            }
+        };
+        Ok(Self::new(
+            fields,
+            FrameSlice::new_subslice(mem, v.as_bytes_ref()),
+        ))
+    }
+}
+
+impl<'frame> Iterator for UdtIterator<'frame> {
+    type Item = Result<
+        (
+            &'frame (String, ColumnType),
+            Option<Option<FrameSlice<'frame>>>,
+        ),
+        ParseError,
+    >;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        // TODO: Should we fail when there are too many fields?
```

Comment:
Let's discuss this. Should we?

If we got more fields than we expected for a UDT, why should we let it silently pass? IMO it deserves at least a warning. Failure is appropriate as well, as I can't find any counterexample.

Reply:
The protocolspec only allows less fields to be sent, and the description implies that it's incorrect to send more:

```
  A UDT value is composed of successive [bytes] values, one for each field of the UDT
  value (in the order defined by the type). A UDT value will generally have one value
  for each field of the type it represents, but it is allowed to have less values than
  the type has fields.
```

The TODO is mostly about considering how to handle a case where the database sends a response that does not conform to the spec. If we don't trust the DB then this method will have to be written a bit differently.